### PR TITLE
Add Poetry support

### PR DESCRIPTION
**Pull request by matthewfeickert**
* Add [Poetry support](https://poetry.eustace.io/)
* Add a Python 3 `requirements.txt`
* Add a Python 3 env Make target
* Add helper setup and run scripts

Poetry and `py3_requirements.txt` are meant to make PR #170 for full Python3 support easier as it gives a workable environment.

As can be seen from the `pyproject.toml` and `poetry.lock` one of the main benefits of Poetry (of many) is to give dependency management that requires **only the true dependancies** to be specified while still achieving the environmental reproducibility that `pip freeze > requirements.txt` offers.

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@justinvasel This is ready for review. Let me know what commends and requested edits you have.

This PR exists only to help our PR #170 so if you want to talk about how they relate I'm happy to if it isn't clear.

**Comment by justinvasel** [comment]
Great, thanks @matthewfeickert! I'll take a look soon and offer any feedback I have. I'm planning to roll out a new feature today that I've been working on (adding user passwords to login credentials), so after that gets merged you may need to make an additional commit, though I've been careful to stay python3-compatible, so maybe not. Will keep you posted.

**Comment by justinvasel** [comment]
Okay, I've just merged a good chunk of changes into the the `master` branch. I hope this doesn't require a lot of extra work on your part here, but you may want to double check that these changes are compatible.

**Comment by matthewfeickert** [comment]
@justinvasel This PR is almost all new files, so while PR #171 will require edits of PR #170 this one is fine. That being said I'll rebase against `origin/master` anyway as that's my habit. But this is ready for review whenever now.

**Comment by matthewfeickert** [comment]
@justinvasel Yeah, this is definitely an issue of what Python version is being called.

> It explicitly installs Poetry with `python3`

True, but all this really means is just that `python3` is running a Python script delivered to it by `curl`ing a web endpoint, not that any sort of environment is setup.

> If I `make python3_env` first and then run `scripts/install_env.sh`, it uses python3 instead of python2 and installs fine

Yes, here you're inside your `env` virtual environment which as it was setup with `python3 -m venv` will return your Python3 installation when

```
$(which python) --version
```

is run inside of it.

So this is clearly an issue of Poetry calling `python` regardless of what environment it is in and so just gets whatever `which python` evaluates to. Let me see if I can figure out a way to force Poetry to look for Python 3 and fail if not present.

**Comment by matthewfeickert** [comment]
@justinvasel Okay, this issue should now be fixed after the rebase that I just added (so you'll have to get the modified versions).

You can test this for yourself in a Docker container by running the following:
```
docker run --rm -it -v $PWD:/home/docker/data -w /home/docker/data atlasml/ml-base:centos
```
and then of course executing

```
bash scripts/install_env.sh
```

Let me know your thoughts.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[tool.poetry]
+name = "whips"
```
Comment (justinvasel) [inline pyproject.toml]: Clarifying question: Do we need to manually increment this number in addition to incrementing the one in `app/__init__.py`, or is there some fancy automagical thing that happens somehow?

Comment (justinvasel) [reply pyproject.toml]: And to clarify my clarifying question, I meant to tag the line below this one, with the WHIPS versioning; must have clicked the wrong line.

**Comment by matthewfeickert** [comment]
> I see now that WHIPS itself becomes a module in this process, it has a `whips.egg-info` directory produced automatically during the install process. At the moment we don't have any plans to make WHIPS available via package repos or anything, so do we need to do anything with it in that case?

Poetry does install whips, but you can see that it doesn't treat it as a dependency as

```
grep whips poetry.lock
```

returns no output. So, no, there is nothing that needs to be done here, but as Poetry is both a "Python packaging and dependency management" tool then it just does this along the way (as a result of [PEP 517](https://www.python.org/dev/peps/pep-0517/)) to make it easy if you _did_ want to distribute it.

> Is it good form to commit these files even if it's not going to get posted to a package repository? 

No, no need to ever version control eggs (I think).

> And if not, is it reasonable to add it to the `.gitignore` list?

Good call. Let me amend that to the PR.

File: pyproject.toml (hunk 1 above)
Comment (matthewfeickert) [reply pyproject.toml]: > Do we need to manually increment this number in addition to incrementing the one in `app/__init__.py`

Yeah, that would be good form.

> is there some fancy automagical thing that happens somehow?

I'm not sure you you're currently doing it, but I would recommend using [`bumpversion`](https://github.com/peritus/bumpversion). Very handy. :+1: 

> And to clarify my clarifying question, I meant to tag the line below this one, with the WHIPS versioning; must have clicked the wrong line.

No, you did everything right. GitHub's comment UX kinda sucks with regards to specifying what line you're actually commenting on. It looks better if you view the files under the "Files changed" tab in the PR as these comments show up there as well.

**Comment by matthewfeickert** [comment]
@justinvasel `pip` eggs are now ignored. :+1:

**Comment by justinvasel** [comment]
Great! I'm happy to merge it in. Thanks for this!

**Pull request merged**